rfc/plugins.md

```diff
@@ -0,0 +1,338 @@
+
+Extension points will be defined and exposed via PF4J.
+
+### PF4J as a Foundation
+[PF4J](https://github.com/pf4j/pf4j) is to be used as the foundation for all plugin functionality. 
+While designed for in-process extensibility, we’ll be using PF4J to define both the contract for in-process plugins, as well as remote plugins. 
+Alternative (co-process, remote) runtimes would also be available but will be covered later.
+
+PF4J has a lot to offer out of the box that we would need to build ourselves, regardless of final implementation. 
+It doesn’t go as far as OSGI, but provides basic (yet extensible) ClassLoader isolation, a simple extension contract that we can use broadly, and a standardized manifest format. 
+PF4J was originally built for creating in-process plugins only, but its own extension points make it feasible to enable remote plugins using the same contract. 
+What runtime a particular plugin uses would be determined by a combination of: 1) The extension point developer allowing remote invocation, and 2) the plugin developer writing the plugin for remote invocation.
+
+In the case of a remote plugin, the plugin developer will have the freedom to choose language, framework, and runtime. 
+For any remote plugin, an in-process counterpart PF4J plugin will be required to enable communication between a service and the plugin. 
+For example, a remote plugin that exposes its integration through a gRPC server would also need a gRPC transport plugin on the service so that communication between the two can be configured correctly. 
+Other transport plugins, such as Docker co-process, Functions, HTTP, and so-on could also be developed.
+
+An extension point looks like the following:
+
+```kotlin
+// Definition of the PluginStage extension point.
+// The annotation defined here also informs the Spinnaker
+// plugin framwework which runtimes are acceptable. This
+// annotation would additionally be used for other metadata
+// Spinnaker needs to know atop whatever PF4J offers out of
+// the box.
+@SpinnakerExtension(allowRemote = true)
+class PipelineStageExtension : ExtensionPoint {
+  // ...
+}
+```
+
+It is highly recommended to read the documentation of PF4J to fully understand the scope of this proposal: This proposal will not go in depth on PF4J internals or taxonomy.
+
+### Front50 as the Source of Truth
+Operators will need the ability to understand, at a glance, what plugins are installed and enabled or disabled, their versions, as well as any other metadata about a plugin. 
+Whether plugins are implemented as in-process or not, Front50 should be used to track and expose a single source of truth of what plugins are installed in any particular service, as well as which ones are enabled or disabled.
+
+The Front50 plugin registry will also be exposed through Gate, and can be consumed by Deck to load the correct plugins on application startup.
```

Comment:
This initially sounded to me like we would depend on Deck to load plugins for other parts of the (Spinnaker) application. Maybe replace "the correct plugins" with "UI plugins"?